<img src="http://graal.ift.ulaval.ca/adrouin/kover.png" height="50" />

Kover is an *out-of-core* implementation of the Set Covering Machine learning algorithm. It compares large sets of wholes genomes and produces highly interpretable models of phenotypes.

# Introduction

Drouin, A., Giguère, S., Déraspe, M., Marchand, M., Tyers, M., Loo, V. G., Bourgault, A. M., Laviolette, F. & Corbeil, J. (2016). Predictive computational phenotyping and biomarker discovery using reference-free genome comparisons. Under Review. [[ Preprint ]](http://biorxiv.org/content/early/2016/03/27/045153)

> The identification of genomic biomarkers is a key step towards improving diagnostic tests and therapies. We present a new reference-free method for this task that relies on a k-mer representation of genomes and a machine learning algorithm that produces intelligible models. The method is computationally scalable and well-suited for whole genome sequencing studies. The method was validated by generating models that predict the antibiotic resistance of C. difficile, M. tuberculosis, P. aeruginosa and S. pneumoniae. We show that the obtained models are accurate and that they highlight biologically relevant biomarkers, while providing insight into the process of antibiotic resistance acquisition. Kover, an efficient implementation of our method, can readily scale to large genomic datasets. It is open-source and can be obtained from http://github.com/aldro61/kover.

# Installation

For installation instructions, see: http://aldro61.github.io/kover/doc_installation.html

# Example

For an example of use, see: http://aldro61.github.io/kover/doc_example.html

# Documentation

The documentation can be found at: http://aldro61.github.io/kover/
